packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4188,6 +4187,62 @@ mixin RenderObjectWithChildMixin<ChildType extends RenderObject> on RenderObject
   }
 }
 
+/// A mixin for [RenderObject] subclasses with a layout callback. The mixin
+/// guarantees the layout callback will be called even if this [RenderObject]
+/// skips doing layout.
+///
+/// A layout callback is a callback that mutates the [RenderObject]'s render
+/// subtree, invoked within an [invokeLayoutCallback] during the [RenderObject]'s
+/// layout process. Sometimes such callbacks involve rebuilding dirty widgets,
+/// especially if the widget configuration depends on the layout process (most
+/// notably, the [LayoutBuilder] widget). Unlike render subtrees, typically all
+/// dirty widgets (even the off-screen ones) in a widget tree must be rebuilt.
+/// This mixin makes sure the layout callback will be invoked (such that dirty
+/// [Element]s are rebuilt) even when an ancestor [RenderObject] chose to skip
+/// laying out this [RenderObject].
+///
+/// Subclasses must not invoke the layout callback directly. Instead, call
+/// `super.performLayout` in the [performLayout] implementation.
+///
+/// See also:
+///
+///  * [LayoutBuilder] and [SliverLayoutBuilder], which use the mixin.
+mixin RenderObjectWithLayoutCallbackMixin on RenderObject {
+  // The initial value of this flag must be set to true to prevent the layout
+  // callback from being called when the tree has never been laid out.
```

Comment:
I don't understand this comment: _needsRebuild doesn't seem to be used as a guard for runLayoutCallback. And also, why do we not want to run the callback during the first time the tree is laid out?

Reply:
This RenderObject will be added to the PipelineOwner's dirty list, but if its ancestors were never laid out we can't run the layout callback because the constraints are not known yet. Changed the class documentation to

> /// A mixin for [RenderObject] subclasses with a layout callback. The mixin
/// guarantees the layout callback will be called even if this [RenderObject]
/// skips doing layout, unless the [RenderObject] has never been laid out and
/// does not have valid [constraints].

Reply:
This new explanation sounds good to me, but you may have forgotten to push the change up to github.